presto-tests/src/test/java/com/facebook/presto/tests/TestGracefulShutdown.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.tests;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.execution.QueryInfo;
+import com.facebook.presto.server.testing.TestingPrestoServer;
+import com.facebook.presto.server.testing.TestingPrestoServer.TestShutdownAction;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.facebook.presto.execution.QueryState.FINISHED;
+import static com.facebook.presto.memory.TestMemoryManager.createQueryRunner;
+import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+
+@Test(singleThreaded = true)
+public class TestGracefulShutdown
+{
+    private static final long SHUTDOWN_TIMEOUT_MILLIS = 600_000;
+    private static final Session TINY_SESSION = testSessionBuilder()
+            .setCatalog("tpch")
+            .setSchema("tiny")
+            .build();
+
+    private final ListeningExecutorService executor = MoreExecutors.listeningDecorator(newCachedThreadPool());
+
+    @Test(timeOut = SHUTDOWN_TIMEOUT_MILLIS)
+    public void testShutdown()
+            throws Exception
+    {
+        ImmutableMap<String, String> properties = ImmutableMap.of("node-scheduler.include-coordinator", "false");
```

Comment:
Reduce the grace period, so that this test doesn't take too long (I'd set it to 10 seconds)